x-pack/plugin/rollup/src/test/java/org/elasticsearch/xpack/downsample/DataStreamTests.java

```diff
@@ -0,0 +1,235 @@
+import org.elasticsearch.xpack.rollup.Rollup;
+import org.hamcrest.Matchers;
+
+import java.io.IOException;
+import java.time.Instant;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.Locale;
+import java.util.concurrent.ExecutionException;
+
+import static org.elasticsearch.cluster.metadata.MetadataIndexTemplateService.DEFAULT_TIMESTAMP_FIELD;
+import static org.hamcrest.Matchers.equalTo;
+
+public class DataStreamTests extends ESSingleNodeTestCase {
+
+    @Override
+    protected Collection<Class<? extends Plugin>> getPlugins() {
+        return List.of(Rollup.class, DataStreamsPlugin.class);
+    }
+
+    public void testDataStreamDownsample() throws ExecutionException, InterruptedException, IOException {
+        // GIVEN
+        final String dataStreamName = randomAlphaOfLength(5).toLowerCase(Locale.ROOT);
+        putComposableIndexTemplate("1", List.of(dataStreamName));
+        client().execute(CreateDataStreamAction.INSTANCE, new CreateDataStreamAction.Request(dataStreamName)).actionGet();
+        indexDocs(dataStreamName, 10, Instant.now().toEpochMilli());
+        final RolloverResponse rolloverResponse = client().admin().indices().rolloverIndex(new RolloverRequest(dataStreamName, null)).get();
+        // NOTE: here we calculate a delay to index documents because the next data stream write index is created with a start time of
+        // (about) two hours in the future. As a result, we need to have documents whose @timestamp is in the future to avoid documents
+        // being indexed in the old data stream backing index.
+        final String newIndexStartTime = client().admin()
+            .indices()
+            .prepareGetSettings(rolloverResponse.getNewIndex())
+            .get()
+            .getSetting(rolloverResponse.getNewIndex(), IndexSettings.TIME_SERIES_START_TIME.getKey());
+        indexDocs(dataStreamName, 10, Instant.parse(newIndexStartTime).toEpochMilli());
+        client().admin()
+            .indices()
+            .updateSettings(
+                new UpdateSettingsRequest().indices(rolloverResponse.getOldIndex())
+                    .settings(Settings.builder().put(IndexMetadata.SETTING_BLOCKS_WRITE, true).build())
+            )
+            .actionGet();
+
+        // WHEN (simulate downsampling as done by an ILM action)
+        final String downsampleTargetIndex = DataStream.BACKING_INDEX_PREFIX + dataStreamName + "-downsample-1h";
+        final DownsampleAction.Request downsampleRequest = new DownsampleAction.Request(
+            rolloverResponse.getOldIndex(),
+            downsampleTargetIndex,
+            new DownsampleConfig(DateHistogramInterval.HOUR)
+        );
+        final AcknowledgedResponse downsampleResponse = client().admin()
+            .indices()
+            .execute(DownsampleAction.INSTANCE, downsampleRequest)
+            .actionGet();
+
+        client().admin()
```

Comment:
Maybe add this is a comment?

Reply:
I was actually trying to see if I can fix the issue...I will spent a couple more hours if I can fix it (maybe, even as part of another review so that I can ask for someone else's review).